--- client.py	2020-10-14 15:57:27.606050110 +0200
+++ client.py	2020-10-14 15:59:36.274509089 +0200
@@ -408,7 +408,7 @@
             conn, command)
         if not response.startswith(const.VERSION):
             raise ClientException('Memcached version failed', response)
-        version, number = response.split()
+        version, number = response.split(maxsplit=1)
         return number
 
     @acquire
